### PR TITLE
Use `dict` to store data for {nan,}arg{min,max} in certain cases

### PR DESCRIPTION
**Pull request by pentschev**
Certain array types don't support structed arrays (e.g., CuPy). In such cases, we may need to resort to a different data type. Details have been presented in https://github.com/dask/dask/issues/7997 and this is an attempt at solving that problem.

With this change, data stored during {nan,}arg{min,max} uses a `dict` instead of a structured array, provided that the array type fails during structured array creation. An additional change to `_concatenate2` is required to handle the case where an attempt is made to concatenate `dict`s.

### CODE REVIEW

**Comment by quasiben** [comment]
Thank you @pentschev for submitting a PR.  We'll need to rebase/merge with main since https://github.com/dask/dask/pull/8013 reorganizes the tests.  

@pentschev can you spend a little time measuring if there is any negative performance impact for the NumPy/CPU case with the PR.  i would be surprised if there is any but confirming would be could.

@rabernat, if you have a moment, I'm curious if you've ever run into issues with structured arrays when using xarray and dask.  If not, apologies for the ping.

Lastly, while this PR generally looks ok to me we should find a third party to review before merging in.  @mrocklin if you have time that would be appreciated, however, we can also find additional parties to review

**Comment by pentschev** [comment]
I don't see any noticeable difference in performance, see plot below:

visualization

The above computes `da.argmin` on 3D arrays of same length in all dimensions (e.g., `(10, 10, 10) .. (1000, 1000, 1000)`) and chunk sizes as half the dimension length in each dimensions, totalling 8 chunks. Each line represents whether is a "Dask main branch" or "PR 8014" run, and the choice of `axis` (`[None, 0, 1, 2]`). Probably not the best way to visualize but gives an idea that time is about the same when comparing both original Dask vs the change in this PR.

The particular setup used above is not based on any pre-existing benchmark but just something I quickly thought of, so if anyone has a specific performance measurement in mind of what they would like to see, please let me know.

**Comment by pentschev** [comment]
> Thanks @pentschev for taking the time to benchmark the changes. There is a planned released for this Friday [dask/community#176](https://github.com/dask/community/issues/176) -- would you be ok to wait until Monday to merge in or do you think this is critical to getting into tomorrow ?

I think it's fine to wait until Monday. Both this and https://github.com/dask/dask/pull/8021 will be required for the sgkit case, so no rush to merge this while the other PR is pending.

File: dask/array/reductions.py [hunk 1]
```diff
@@ -1037,10 +1065,18 @@ def arg_reduction(x, chunk, combine, agg, axis=None, split_every=None, out=None)
         ((name,) + k, (chunk, (old,) + k, axis, off))
         for (k, off) in zip(keys, offset_info)
     )
-    # The dtype of `tmp` doesn't actually matter, just need to provide something
+
+    dtype = np.argmin(asarray_safe([1], like=meta_from_array(x)))
+    meta = None
+    if is_arraylike(dtype):
+        # This case occurs on non-NumPy types (e.g., CuPy), where the returned
+        # value is an ndarray rather than a scalar.
```
[the remark this answers is not in the document]
Comment (jsignell) [reply dask/array/reductions.py:1073]: Ok cool cool. I misunderstood.

**Comment by tomwhite** [comment]
> @tomwhite you might be interested in this PR as it has come it in trying to accelerate some cupy-sgkit workloads. If you have comment that would be extremely appreciated

Thanks for the heads up @quasiben. Is there a sgkit workload that you had in mind? We use `np.unique` (with `return_inverse`, `return_counts`, etc) in some places, so the changes here should be useful for that.

cc @eric-czech

**Comment by pentschev** [comment]
@tomwhite the particular workflow is `regenie`, specifically this PR is necessary for https://github.com/pystatgen/sgkit/blob/5d58d09f22f023c866e6b252662f84f83302eb61/sgkit/stats/regenie.py#L417 .

Also the `np.unique` case is used in https://github.com/pystatgen/sgkit/blob/5d58d09f22f023c866e6b252662f84f83302eb61/sgkit/stats/regenie.py#L471, although neither `return_inverse` nor `return_counts` are used there. Thus, I also have a PR to treat that one particular case in https://github.com/dask/dask/pull/8021. In the future we will hopefully deal with `np.unique` in a better way (see https://github.com/dask/dask/issues/2851), but this discussion is better left for https://github.com/dask/dask/pull/8021.

Also just to let you know, I have already local changes to support Dask+CuPy for `regenie` (requires both Dask PRs I mentioned above), we're only working with NVIDIA's legal team to get approval to contribute to sgkit, so hopefully next week I'll open a PR for that.

**Comment by quasiben** [comment]
If folks don't have any objection I'll plan to merge in later this afternoon.

**Comment by quasiben** [comment]
Thanks again @pentschev

**Pull request merged**